### PR TITLE
Large-scale refactor of scan function

### PR DESCRIPTION
**Pull request by Nexarian**
# Summary
- Returns a list of the discovered devices for use as a utility function.
- Ability to silence output for use as a utility.
- Improve performance of multi-threaded scan by using a Queue.
- Switch to "f" strings for interactive output.
- General code flow improvements and encapsulation.
- Switch to more generalized wording. Not all Tesla Energy Gateways are Powerwalls anymore. Add ability to work with standalone inverters.

# Rationale
I am building a utility that I plan to run periodically to collect string data from my standalone Tesla inverters. Because this requires local LAN IP access, there is a possibility that  the IP address will change on the LAN (especially if it's on wifi). Therefore, my utility should do a quick scan of the network every time it reports data. Thus, I needed something that would report the list of discovered devices and do it quickly, hence the emphasis on scanning as a utility and multi-threading improvements.

This sucked me into a wholesale refactor of this module, pretty much no line was left untouched. The style is rather different than the rest of pypowerwall. If that's not OK I'm happy to work on changing things to match what is preferred in the repo, though I generally believe the refactored code is tighter, easier to read, and more resilient to errors.

# Testing

See the code below for using this as a utility. The call is the very last line, in main.py.

The `manage_ip_route` function is so useful that we should probably put it in the README. I'll cut a PR for that next.

<details>
<summary>File: manual_reporter.py</summary>

```
import json
import socket
from enum import Enum
from typing import Any, Dict, Final

import pypowerwall
from pyroute2 import IPRoute, NetlinkError


class RouteOperation(Enum):
    ADD = 'add'
    DELETE = 'del'

def manage_ip_route(operation: RouteOperation, destination: str, gateway: str, interface: str = None):
    """
    Manages an IP route using pyroute2's IPRoute, with 'onlink' handling.

    Parameters:
    - operation: RouteOperation.ADD or RouteOperation.DELETE
    - destination: The network or IP address in CIDR notation (e.g., "192.168.1.0/24")
    - gateway: The gateway IP address (e.g., "192.168.1.1")
    - interface: The optional network interface (e.g., "eth0").
                 If not provided, the route is managed without specifying an interface.

    Example usage:
    manage_ip_route(RouteOperation.ADD, "192.168.1.0/24", "192.168.1.1")
    manage_ip_route(RouteOperation.DELETE, "192.168.1.0/24", "192.168.1.1", "eth0")
    """
    ip: Final = IPRoute()

    # Build route parameters
    route_params = {
        'family': socket.AF_INET6 if ':' in destination else socket.AF_INET,
        'dst': destination,
        'gateway': gateway
    }

    if operation == RouteOperation.ADD:
        route_params['flags'] = ['onlink']

    try:
        # Lookup interface index if interface is specified
        if interface:
            idxs = ip.link_lookup(ifname=interface)
            if not idxs:
                print(f"Interface '{interface}' not found.")
                return
            route_params['oif'] = idxs[0]
        # Perform the route operation
        if operation == RouteOperation.ADD:
            ip.route('add', **route_params)
            action = "added"
        elif operation == RouteOperation.DELETE:
            ip.route('del', **route_params)
            action = "deleted"

        print(f"Route {action}: {destination} via {gateway}" + (f" dev {interface}" if interface else "") + (f" {','.join(route_params['flags'])}" if 'flags' in route_params else ""))
    except NetlinkError as e:
        # Handle netlink errors
        action = "adding" if operation == RouteOperation.ADD else "deleting"
        print(f"Error {action} route: {e}")
    except Exception as e:
        # Handle other exceptions
        print(f"An unexpected error occurred: {e}")
    finally:
        # Ensure resources are cleaned up
        ip.close()

#TEDAPI
def TEDAPI_REPORT(gw_pw: str, device: Dict[Any, Any]) -> None:
    email: Final = "email@domain.com"
    timezone: Final = "[TIMEZONE]"
    host: Final = "192.168.91.1"
    password: Final = ""
    pw: Final = pypowerwall.Powerwall(host,password,email,timezone,gw_pwd=gw_pw)

    # Some System Info
    print(f"Site Name: {pw.site_name()} - Firmware: {pw.version()} - DIN: {pw.din()}")
    print(f"System Uptime: {device['up_time'] if 'up_time' in device else pw.uptime()}\n")

    # Display String Data
    string_data = pw.strings()
    print("String Data: %s\n" % json.dumps(string_data, indent=4))
    power = 0
    for value in string_data.values():
        power = power + value["Power"]
    print("String Power Total:", power)

    # Display Device Vitals
    vitals = pw.vitals()
    print("Vitals Data: %s\n" % json.dumps(vitals, indent=4))
```
</details>

<details>
<summary>File: main.py</summary>

```
import sys
import os
import sys
from typing import Any, Dict, Final, List

from pypowerwall import scan

from manual_reporter.manual_reporter import (TEDAPI_REPORT, RouteOperation,
                                             manage_ip_route)

passwords: Dict[str, str] = {
    '[DEVICE DIN 1]': "[GATEWAY PASSWORD 1]",
    '[DEVICE DIN 2]': "[GATEWAY PASSWORD 2]"
}

def main(devices: List[Any]) -> int:
    ethernet_interface: Final[str] = "eth0"
    tedapi_subnet: Final[str] = "192.168.91.0/24"

    for device in devices:
        din: Final[str] = device.get('din', '')
        ip: Final[str] = device.get('ip', '')
        if not din or not ip:
            continue

        password: Final[str] = passwords.get(din, '')
        if not password:
            print(f"Password not found for device with DIN {din}. Skipping")
            continue

        try:
            manage_ip_route(RouteOperation.ADD, tedapi_subnet, ip, ethernet_interface)
            TEDAPI_REPORT(password, device)
        except Exception as e:
            print(f"Error manging route or reporting for device {din}: {e}")
        finally:
            manage_ip_route(RouteOperation.DELETE, tedapi_subnet, ip, ethernet_interface)

def is_root() -> bool:
    return os.geteuid() == 0

if __name__ == '__main__':
    if not is_root():
        # Re-run the script with sudo
        print("This script needs to run with administrative privileges. Requesting sudo...")
        try:
            os.execvp("sudo", ["sudo", "python3", sys.argv])
        except Exception as e:
            print(f"Failed to elevate privileges: {e}")
            sys.exit(1)

    main(scan.scan(max_threads=256))
```
</details>

# Credit
Some credit goes to ChatGPT, which helped with some of the refactoring.

### CODE REVIEW

**Comment by jasonacox** [comment]
Thanks for this @Nexarian !- I like the update.

I seem some issues scanning my Powerwall 2 and +.  Here is a quick scan (notice what is scanned doesn't match what is ultimately found) also the "unknown" which seems to be what the gateway is sending back, oddly.

Here is the OLD code:

image

Here is the PR code:

image

So we need to address a few issues here:

* Need to fix the interactive scan IP that is showing when a TEG is found so it matches the reported.
* The "Unknown" is likely from the `teg_type` payload, we may need to switch to `device_type` instead.
* Minor color issue with "Code" for found items

**Comment by Nexarian** [comment]
@jasonacox Updated with what I believe fixes for all the issues you've mentioned.

**Comment by Nexarian** [comment]
On mine, `device_type` is simple `teg` which is less useful than `Powerwall` so I've kept it `Powerwall` for now.

**Comment by jasonacox** [comment]
Fixed! Thanks @Nexarian - I'll add some a version bump and release notes.

File: pypowerwall/scan.py [hunk 1]
```diff
@@ -8,200 +8,294 @@
...
+        # Expected response from PW3
+        # {"code":403,"error":"Unable to GET to resource","message":"User does not have adequate access rights"}
+        if "User does not have adequate access rights" in response_pw3.text:
+            # Found PW3
+            if context.interactive:
+                print(f"{host} OPEN{context.dim()} - {context.subbold()}Found Powerwall 3 [Supported in Cloud Mode only]{context.normal}")
+            result_queue.put({
+                'ip': addr,
+                'din': 'Powerwall-3',
+                'firmware': 'Supported in Cloud Mode only - See https://tinyurl.com/pw3support'
```
Comment (jasonacox) [inline pypowerwall/scan.py:190]: Since we are updating, I'll update this to include insturctions that PW3 users can use cloud or TEDAPI mode.

**Comment by jasonacox** [comment]
Ready to merge and release update - @Nexarian let me know if this work for you.

**Comment by Nexarian** [comment]
@

> Ready to merge and release update - @Nexarian let me know if this work for you.

Looks good to me! Please merge!

**Comment by jasonacox** [comment]
Merged! View at:
https://pypi.org/project/pypowerwall/0.12.1/

**Pull request merged**